oauth2_http/java/com/google/auth/oauth2/CertificateIdentityPoolSubjectTokenSupplier.java

```diff
@@ -0,0 +1,122 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.auth.oauth2;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonPrimitive;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.InvalidPathException;
+import java.nio.file.Paths;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.Base64;
+
+/**
+ * Provider for retrieving subject tokens for {@link IdentityPoolCredentials} by reading an X.509
+ * certificate from the filesystem. The certificate file (e.g., PEM or DER encoded) is read, the
+ * leaf certificate is base64-encoded (DER format), wrapped in a JSON array, and used as the subject
+ * token for STS exchange.
+ */
+public class CertificateIdentityPoolSubjectTokenSupplier
+    implements IdentityPoolSubjectTokenSupplier {
+
+  private static final Gson GSON = new Gson();
+  private final IdentityPoolCredentialSource credentialSource;
+
+  CertificateIdentityPoolSubjectTokenSupplier(IdentityPoolCredentialSource credentialSource) {
+    this.credentialSource = checkNotNull(credentialSource, "credentialSource cannot be null");
+    // This check ensures that the credential source was intended for certificate usage.
+    // IdentityPoolCredentials logic should guarantee credentialLocation is set in this case.
+    checkNotNull(
+        credentialSource.certificateConfig,
+        "credentialSource.certificateConfig cannot be null when creating"
+            + " CertificateIdentityPoolSubjectTokenSupplier");
+  }
+
+  private static X509Certificate loadLeafCertificate(String path)
+      throws IOException, CertificateException {
+    byte[] leafCertBytes;
+    try {
+      // IdentityPoolCredentials should have already validated the path exists via X509Provider.
+      leafCertBytes = Files.readAllBytes(Paths.get(path));
+    } catch (InvalidPathException e) {
+      throw new IOException("Invalid certificate file path provided: " + path, e);
```

Comment:
qq, why is this re-thrown as an IOException?

Reply:
We catch InvalidPathException and re-throw it as an IOException primarily because if `loadLeafCertificate()` throws InvalidPathException directly, the `getSubjectToken()` method (which calls it) is required by the `IdentityPoolSubjectTokenSupplier` interface to only throw IOException for checked exceptions.
So, we'd have to catch the InvalidPathException in getSubjectToken anyway and wrap it in an IOException there. Catching and wrapping it inside `loadLeafCertificate()` keeps the error handling closer to the source of the potential path issue and simplifies `getSubjectToken`.

Reply:
> So, we'd have to catch the InvalidPathException in getSubjectToken anyway and wrap it in an IOException there.

I believe `InvalidPathException` is a a runtime exception and you won't need to add the exception to the method signature. Since IOExceptions are not being explicitly caught and handled, I think `InvalidPathException` should can also just be thrown at runtime.